src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -754,106 +759,247 @@ public static boolean bothSingleByteOptimizable(DynamicObject string, DynamicObj
 
-        @Child private EncodingNodes.CheckEncodingNode checkEncodingNode = EncodingNodesFactory.CheckEncodingNodeGen.create(null, null);
         @Child private ToStrNode toStr = ToStrNodeGen.create(null);
+        @Child private CountRopesNode countRopesNode = CountRopesNode.create();
+
+        @Specialization(guards = "args.length == size", limit = "getCacheLimit()")
+        public int count(VirtualFrame frame, DynamicObject string, Object[] args,
+                @Cached("args.length") int size) {
+            final Rope[] ropes = argRopes(frame, args);
+            return countRopesNode.executeCount(string, ropes);
+        }
+
+        @Specialization(replaces = "count")
+        public int countSlow(VirtualFrame frame, DynamicObject string, Object[] args) {
+            final Rope[] ropes = argRopesSlow(frame, args);
+            return countRopesNode.executeCount(string, ropes);
+        }
+
+        @ExplodeLoop
+        protected Rope[] argRopes(VirtualFrame frame, Object[] args) {
+            final Rope[] strs = new Rope[args.length];
+            for (int i = 0; i < args.length; i++) {
+                strs[i] = rope(toStr.executeToStr(frame, args[i]));
+            }
+            return strs;
+        }
+
+        protected Rope[] argRopesSlow(VirtualFrame frame, Object[] args) {
+            final Rope[] strs = new Rope[args.length];
+            for (int i = 0; i < args.length; i++) {
+                strs[i] = rope(toStr.executeToStr(frame, args[i]));
+            }
+            return strs;
+        }
+
+        protected int getCacheLimit() {
+            return getContext().getOptions().DEFAULT_CACHE;
+        }
+    }
+
+    @ImportStatic({ StringGuards.class, StringOperations.class })
+    public abstract static class CountRopesNode extends TrTableNode {
+
+        public static CountRopesNode create() {
+            return CountRopesNodeFactory.create(null);
+        }
+
+        public abstract int executeCount(DynamicObject string, Rope[] ropes);
 
         @Specialization(guards = "isEmpty(string)")
         public int count(DynamicObject string, Object[] args) {
             return 0;
         }
 
+        @Specialization(guards = { "!isEmpty(string)", "args.length > 0", "cachedArgs.length == args.length",
+                "argsMatch(cachedArgs, args)", "encodingsMatch(string, cachedEncoding)"
+        })
+        public int countFast(DynamicObject string, Rope[] args,
+                @Cached(value = "args", dimensions = 1) Rope[] cachedArgs,
+                @Cached("encoding(string)") Encoding cachedEncoding,
+                @Cached(value = "makeSqueeze()", dimensions = 1) boolean[] squeeze,
+                @Cached("findEncoding(string, cachedArgs)") Encoding compatEncoding,
+                @Cached("makeTables(string, cachedArgs, squeeze, compatEncoding)") TrTables tables) {
+            return processStr(string, squeeze, compatEncoding, tables);
```

Comment:
At this point of caching, it might be worth just having the count loop in here if compatEncdoing.isAsciiCompatible().